### DISCUSSION TITLE
Shouldn't thumbnails be generated in another cache directory?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1406] [Q&A]
I noticed that thumbnails are created in the same directory as the file.  It seems to me that the derived files like this should be created in a separate folder or at least a temporary space.  We might want to store the thumbnails in `/tmp` or `/var/tmp` or just another directory that won't interfere with being able to backup our data files.

**Comment by ganigeorgiev** [comment]
Thumbs are part of the storage directory and generated lazily on first access. If we move them in a system tmp directory, I'm not sure if we can guarantee that they will not be deleted automatically by the OS or by some other process (and it'll hurt the overall performance if we have to pregenerate them when hundreds of images are requested at the same time).

I don't think it is a bad thing to be part of the `pb_data` directory and its backup.

**Reply by dschissler** [reply]
Could we gain more control over thumbnails being generated?  I don't know what is so special about 100x100 that it needs to automatically be created.  Not having control over that means in certain use cases that it's just an inefficient data storage.

Could you put this on a roadmap to have better controls over it?

**Reply by ganigeorgiev** [reply]
`100x100` is what is used in the Admin UI.

> Not having control over that means in certain use cases that it's just an inefficient data storage

That's a little exaggerated. You can provide your own app specific thumb sizes from the `file` field options.

> Could you put this on a roadmap to have better controls over it?

Better control for what? Currently we support both various thumb sizes (including preserving the aspect ratio) and crop anchors - https://pocketbase.io/docs/files-handling/#file-url.

**Reply by dschissler** [reply]
> That's a little exaggerated. You can provide your own app specific thumb sizes from the file field options.

but what about disabling thumbnails altogether?  If the use case is storing images that don't need thumbnails and then thumbnails are automatically generated, then to me that classifies as inefficient.  It's a great feature, but if it's not wanted then it's just junk data.

**Reply by ganigeorgiev** [reply]
You'll always have at least one thumbnail used in the Admin UI.

> but what about disabling thumbnails altogether?

But why you'll want to do that? I'm ok with adding knobs and features but without a context why and how this will be used I don't want to bloat the platform.

Personally I don't consider a single 100x100 thumb "inefficient". Is that really a bottleneck for your use case?

**Reply by dschissler** [reply]
It's not a bottleneck for my use case now.

But there are things to consider.  What if the images being stored are already pretty small and someone wants to store a hundred thousand of them?  If the images are already tiny then the thumbnails are a huge percentage of the original file size.

Just something to think about for post 1.0 stuff as more people use the system.

**Reply by LucasGabrielBravo** [reply]
Note: Translated by Google translator

Hello, I really like Pocketbase, I think it's perfect for many situations, but here's a case where thumbnails are undesirable.

I made a site for a friend where in PocketBase there is a "collection" of "Performed events" in addition to the data of each event, the collection has a field for the images of the event and the user can put as many images as he wants. The problem is that there are many events and each event has several images. Just by opening the collection in PocketBase Admin the processing and memory cost skyrockets to more than 10x. I'm not sure, but I think this is due to the creation of the thumbnails. I don't care about the size they occupy on disk, but if the processing and memory consumption is really affected due to the creation of thumbnails, I think they should be optional.
Currently the site is made with NextJS. PocketBase and the website are running on Docker on a VPS with 1Gb of RAM and whenever I open PocketBase Admin the VPS usage goes to 100%

**Reply by ganigeorgiev** [reply]
@LucasGabrielBravo The thumbnails are generated lazily **only on the first access**. While the thumbnail generation could be memory intensive depending on the resolution of the images, this should be only one time operation.

I'm not able to reproduce the issue locally when browsing the Admin UI.

If you are experiencing high memory consumption every time when you load your website, then most likely this is a Next.js issue (especially with SSR).

If you are still not able to identify the issue, please file a new issue/discussion with **minimal** reproducible repo and I'll try to investigate it.

**Reply by LucasGabrielBravo** [reply]
Thanks for the quick response.

The high memory consumption happens when I open the PocketBase administrator interface (perhaps it wasn't very clear due to the translation by google translator) precisely in the collection that has the images and it always has this behavior of excessive consumption, not just in the first time. One thing I just noticed is when I disable the visualization of the "images" column, consumption is normal and when I enable it, it consumes a lot of memory again. I imagine that when disabling the "images" column and shortly after enabling it again, a recreation of the thumbnails should happen. Anyway, even with the thumbnails already created whenever I open PB Admin apparently it tries to recreate. But I will ask my friend to reduce the size of the images to see if the problem persists. I'm also going to create an abient just to test these image thumbnail issues and if that's the case I'll open a new discussion.

I also leave some images of the VPS console

dfc9f8a5126f - PocketBase
aebcb92623cf - NextJS

When accessing the collection with the hidden image column
![](https://i.imgur.com/U1IjHbp.png)

When accessing the collection with the column of images visible
![](https://i.imgur.com/35gFsnp.png)
![](https://i.imgur.com/T2Yk8ue.png)

**Reply by ganigeorgiev** [reply]
> whenever I open PB Admin apparently it tries to recreate

I don't think this is the case.

What version of PocketBase are you using?

The memory consumption could be also a result of serving the file to the client and it accumulates because the GC doesn't release it immediately.

You can set a soft-memory limit to trigger the GC more frequently with the `GOMEMLIMIT` env variable (eg. `GOMEMLIMIT=300MiB`).

Another thing I guess that could trigger high memory consumption could be also if your images are stored in a S3 storage since in this scenario we are sending S3 API request for each requested file.

I don't think there is much we can do to minimize the memory consumption when serving files, but I'll try to find time to investigate it this weekend.

**Reply by ganigeorgiev** [reply]
@LucasGabrielBravo 

I've tried to reproduce the issue by creating a sample collection with 100 records each with 25 images.

My first test was with directly access the Admin UI.

By looking at the `allocs` and the `profile` pprof graphs after listing and loading the thumbs, I didn't noticed anything  out of the ordinary.

The temp CPU spikes on all cores are expected due to the large number of fired goroutines for each file request. The Admin UI shows 30 items on the first page each with 25 images, resulting in 750 concurrent requests (_the browser do some batching on its own but still the requests are fired within couple ms difference_).

I've also performed another test and added `return nil` statement at the top of the file serving endpoint to rule out any eventual files serve and processing overhead and the CPU usage was more-or-less similar to the previous test.

Locally, the memory usage was within max ~200-300MB and I think this is normal because it is just bunch of accumulated memory from the general request processing (parsing the request data, generating activity log, etc.) and the GC will clean it up when there's a need for releasing the resources.

If your VPS doesn't have much resources, you can experiment with the [`GOMEMLIMIT` and the `GOMAXPROCS` env variables](https://pkg.go.dev/runtime#hdr-Environment_Variables).

One thing that we can optimize in the Admin UI in order to limit the number of file requests is to show only an "excerpt" of the uploaded files, eg. the first 15 per record (_we do something similar with the multiple `relation` values_), since currently if the user upload 100 images in 1 record, we'll display in the record listing row all 100 thumbs, which doesn't look very good and it is not very useful. I'll apply the minor optimization in the next v0.15.0 release.

**Reply by LucasGabrielBravo** [reply]
Hello! Thanks so much for taking the time to look into this issue. I thought the idea of displaying only 15 thumbnails was very good and will be very welcome.

Apparently the problem with memory consumption and processing happened because of the size of the images, but I intend to do more tests to be sure.

Reducing the size of the images and leaving the "images" column hidden seems to help a lot with usability.

I look forward to 0.15 and thank you again for your attention.
